File: apps/ios/ExpCore/Sources/Domain/AgentUsagePresentation.swift
import Foundation

/// EXP-484: how a machine's per-agent auth status and rate-limit usage
/// present.
///
/// The device is the only writer: it probes its agent CLIs locally and reports
/// `agentAccounts` / `agentUsage` on register/heartbeat, and the synced
/// `devices` row is what every client draws. Nothing here fetches, refreshes
/// or holds a credential — this is read-only visibility.
///
/// Pure and hand-mirrored ×4 (web `lib/agent-usage.ts`, Android
/// `domain/AgentUsagePresentation.kt`, desktop `ui/src/usage_bar.rs`) against
/// the SAME fixture, so a bar reads identically on every client. Change a rule
/// here and the other three move with it.

/// How loud a usage percentage reads. Thresholds are locked ×4: ≥95 danger,
/// ≥75 warning, anything below normal.
public enum AgentUsageSeverity: String, Equatable, Sendable {
    case normal
    case warning
    case danger
}

/// One session's live usage: the agent it runs and that agent's report.
public struct SessionAgentUsage: Equatable, Sendable {
    public let agent: String
    public let usage: AgentUsage

    public init(agent: String, usage: AgentUsage) {
        self.agent = agent
        self.usage = usage
    }
}

/// One usage card: a titled bar with its percentage, tone and caption. Pure
/// presentation — `AgentUsagePresentation.usageGroups` is the only maker.
public struct UsageCard: Equatable, Sendable, Identifiable {
    /// The wire window key the card came from (`session`, `weekly`,
    /// `model:fable`, `credits`, …).
    public let key: String
    public let title: String
    /// Nil when the machine reported the window but no number for it.
    public let percent: Double?
    public let severity: AgentUsageSeverity
    /// Empty when there is nothing to say under the bar.
    public let caption: String

    public var id: String { key }

    public init(key: String, title: String, percent: Double?, severity: AgentUsageSeverity, caption: String) {
        self.key = key
        self.title = title
        self.percent = percent
        self.severity = severity
        self.caption = caption
    }
}

/// A titled run of cards. `key` is stable (`session` / `weekly` / `other`) so
/// a view can special-case one without matching on its title.
public struct UsageGroup: Equatable, Sendable, Identifiable {
    public let key: String
    public let title: String
    public let cards: [UsageCard]

    public var id: String { key }

    public init(key: String, title: String, cards: [UsageCard]) {
        self.key = key
        self.title = title
        self.cards = cards
    }

    init(key: String, title: String, windows: [AgentUsageWindow], now: Date) {
        self.init(
            key: key,
            title: title,
            cards: windows.map { window in
                UsageCard(
                    key: window.key,
                    title: AgentUsagePresentation.cardTitle(window),
                    percent: window.percent,
                    severity: AgentUsagePresentation.severity(window.percent),
                    caption: AgentUsagePresentation.cardCaption(window, now: now)
                )
            }
        )
    }
}

/// EXP-804: `coding_sessions.blocked` — the agent's usage wall as row state.
/// Every field is optional for the same reason the server's zod mirror is
/// `.nullish()` throughout: a newer device naming a window this build has no
/// name for must degrade that field, never fail the whole decode and leave a
/// walled run rendering healthy.
public struct CodingSessionBlocked: Decodable, Equatable, Sendable {
    public let kind: String?
    public let agent: String?
    public let window: String?
    public let resetsAt: String?
    public let since: String?

    public init(
        kind: String? = nil,
        agent: String? = nil,
        window: String? = nil,
        resetsAt: String? = nil,
        since: String? = nil
    ) {
        self.kind = kind
        self.agent = agent
        self.window = window
        self.resetsAt = resetsAt
        self.since = since
    }
}

public enum AgentUsagePresentation {
    /// Numbers older than this are not current enough to draw. Locked ×4.
    public static let freshWindow: TimeInterval = 15 * 60

    // MARK: - Parsing

    /// One agent's usage report from the stored jsonb string. Nil on absent or
    /// unparsable JSON — usage is simply not shown then, never guessed. A
    /// single malformed WINDOW is dropped rather than blanking the report
    /// (`AgentUsageWindow.init(from:)`, which also clamps the percentage).
    public static func parse(_ json: String?) -> AgentUsage? {
        guard let json, let data = jsonData(json) else { return nil }
        return try? JSONDecoder().decode(AgentUsage.self, from: data)
    }

    /// The whole `agentUsage` map (agent id → report) from the stored jsonb.
    public static func parseMap(_ json: String?) -> [String: AgentUsage]? {
        guard let json, let data = jsonData(json) else { return nil }
        return try? JSONDecoder().decode([String: AgentUsage].self, from: data)
    }

    /// The `agentAccounts` map (agent id → sign-in status) from the stored jsonb.
    public static func parseAccounts(_ json: String?) -> [String: AgentAccount]? {
        guard let json, let data = jsonData(json) else { return nil }
        return try? JSONDecoder().decode([String: AgentAccount].self, from: data)
    }

    /// EXP-804: the run's usage wall from the stored jsonb string. Nil on
    /// absent or unparsable JSON — a run is then simply not shown as blocked,
    /// never guessed.
    public static func parseBlocked(_ json: String?) -> CodingSessionBlocked? {
        guard let json, let data = jsonData(json) else { return nil }
        return try? JSONDecoder().decode(CodingSessionBlocked.self, from: data)
    }

    private static func jsonData(_ json: String) -> Data? {
        let trimmed = json.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        return Data(trimmed.utf8)
    }

    // MARK: - Grouping (EXP-688)

    /// The reported windows as the CARDS a usage surface draws, in three fixed
    /// groups. There is no pinned/tracked window any more (EXP-688 deleted the
    /// concept ×4): every window the machine reported is shown, grouped the
    /// way Claude's own app groups them.
    ///
    /// Locked ×4 (web `usageGroups`, Android `usageGroups`, desktop
    /// `usage_groups`) against the same fixture:
    ///   - `session` → "Current session" / card title "Current session";
    ///   - `weekly` + `model:*` → NO title (EXP-694: the card titles already
    ///     say "All models" / "<Label> only", so the group header was noise),
    ///     the all-models window first;
    ///   - anything else (`credits`, codex's `43200`) → "Other" with the wire
    ///     label, in report order.
    /// Empty groups are omitted and the group order never varies. A group with
    /// an EMPTY title draws no header — every renderer skips it.
    public static func usageGroups(_ usage: AgentUsage, now: Date = Date()) -> [UsageGroup] {
        let windows = usage.windows ?? []
        let session = windows.filter { $0.key == sessionWindowKey }
        let weekly = windows.filter { $0.key == weeklyWindowKey }
            + windows.filter { $0.key.hasPrefix(modelWindowPrefix) }
        let other = windows.filter {
            $0.key != sessionWindowKey && $0.key != weeklyWindowKey
                && !$0.key.hasPrefix(modelWindowPrefix)
        }
        return [
            UsageGroup(key: "session", title: "Current session", windows: session, now: now),
            UsageGroup(key: "weekly", title: "", windows: weekly, now: now),
            UsageGroup(key: "other", title: "Other", windows: other, now: now),
        ].filter { !$0.cards.isEmpty }
    }

    // MARK: - Session context + spend (EXP-746)

    /// The heading over the run's own context/spend block in the Usage sheet.
    /// A SIBLING of `usageGroups`, never a group inside it: those cards are
    /// fixture-locked ×4 percentages of a rate-limit window, and a token count
    /// with no percent would draw an empty rail there.
    /// Byte-identical ×4 (web `agent-usage.ts` `CONTEXT_SECTION_TITLE`,
    /// Android `AgentUsagePresentation.kt`, desktop `ui/src/usage_bar.rs`).
    public static let contextSectionTitle = "Context"

    /// The run's context line — `"124k / 200k (62%)"`. Counts round to whole
    /// thousands from 1000 up, the percentage is FLOORED, and an unknown size
    /// prints nothing at all. Locked ×4 by the test
    /// `context usage reads used over size with a percent`.
    public static func formatContextUsage(used: Int, size: Int) -> String? {
        guard size > 0 else { return nil }
        let percent = min(100, max(0, Int((Double(max(0, used)) * 100 / Double(size)).rounded(.down))))
        return "\(formatTokens(max(0, used))) / \(formatTokens(size)) (\(percent)%)"
    }

    /// The run's spend — `"$1.24"`, two decimals. Nil under half a cent (and
    /// for an absent figure): an agent that reports `0.0009` has effectively
    /// spent nothing and "$0.00" reads as a bug. Locked ×4 by the test
    /// `a cost under half a cent renders nothing`.
    public static func formatUsageCost(_ cost: Double?) -> String? {
        guard let cost, cost >= 0.005 else { return nil }
        return String(format: "$%.2f", cost)
    }

    /// Whole thousands from 1000 up (`124000` → `124k`), the raw count below.
    private static func formatTokens(_ count: Int) -> String {
        guard count >= 1000 else { return "\(count)" }
        return "\(Int((Double(count) / 1000).rounded()))k"
    }

    /// The idle-session line Claude's own app shows: a session window at 0%
    /// with nothing to reset has not started yet. Locked ×4.
    public static let sessionNotStartedCaption = "Starts when a message is sent"

    static let sessionWindowKey = "session"
    static let weeklyWindowKey = "weekly"
    static let modelWindowPrefix = "model:"

    /// One card's title: the two named windows read as words, a per-model
    /// window as "<Label> only", and anything else keeps the wire label.
    static func cardTitle(_ window: AgentUsageWindow) -> String {
        if window.key == sessionWindowKey { return "Current session" }
        if window.key == weeklyWindowKey { return "All models" }
        if window.key.hasPrefix(modelWindowPrefix) { return "\(window.label) only" }
        return window.label
    }

    /// The line under a card's bar: the countdown when the window resets, the
    /// not-started sentence for an untouched session window, else nothing.
    static func cardCaption(_ window: AgentUsageWindow, now: Date) -> String {
        if let countdown = resetCountdown(resetsAt: window.resetsAt, now: now) { return countdown }
        if window.key == sessionWindowKey, window.percent == 0 { return sessionNotStartedCaption }
        return ""
    }

    /// Locked thresholds. An unreported percentage reads normal — absence is
    /// not an alarm.
    public static func severity(_ percent: Double?) -> AgentUsageSeverity {
        guard let percent else { return .normal }
        if percent >= 95 { return .danger }
        if percent >= 75 { return .warning }
        return .normal
    }

    // MARK: - Freshness

    /// Whether numbers fetched at [fetchedAt] are current enough to draw.
    /// FAIL-CLOSED like device liveness: an absent or unparsable stamp reads
    /// stale, because claiming an old percentage is current is the bad failure.
    /// A stamp in the future is clock skew, not staleness.
    public static func isFresh(fetchedAt: String?, now: Date = Date()) -> Bool {
        guard let fetchedAt, let fetched = WireTimestamps.parse(fetchedAt) else { return false }
        return now.timeIntervalSince(fetched) < freshWindow
    }

    /// `resets in 2h 10m` / `resets in 3d 14h` / `resets in 45m` / `resets
    /// soon` (under a minute, or already past). Nil when the window never
    /// resets or the stamp is unreadable — the strings are locked ×4, and a
    /// zero smaller unit is dropped (`resets in 2h`, never `resets in 2h 0m`).
    public static func resetCountdown(resetsAt: String?, now: Date = Date()) -> String? {
        guard let resetsAt, let reset = WireTimestamps.parse(resetsAt) else { return nil }
        let minutes = Int(reset.timeIntervalSince(now) / 60)
        guard minutes >= 1 else { return "resets soon" }
        let days = minutes / (60 * 24)
        let hours = (minutes / 60) % 24
        if days > 0 {
            return hours > 0 ? "resets in \(days)d \(hours)h" : "resets in \(days)d"
        }
        if hours > 0 {
            let rest = minutes % 60
            return rest > 0 ? "resets in \(hours)h \(rest)m" : "resets in \(hours)h"
        }
        return "resets in \(minutes)m"
    }

    /// EXP-804: the one-line badge for a run's usage wall — `Rate limited ·
    /// resets in 2h`, or bare `Rate limited` when the agent named no reset
    /// time. Nil when the run is not blocked.
    ///
    /// The wall is ORTHOGONAL to the session state: a blocked run still reads
    /// `running`, so this NEVER replaces `CodingSessionDisplay`'s state — it
    /// renders beside it. An unrecognised `kind` still gets a badge
    /// (`Blocked`): a future device reporting a wall this build has no name
    /// for must not render silent. Locked ×4 (web `blockedBadgeLabel`).
    public static func blockedBadgeLabel(
        _ blocked: CodingSessionBlocked?,
        now: Date = Date()
    ) -> String? {
        guard let blocked else { return nil }
        let label = (blocked.kind ?? "rate_limit") == "rate_limit"
            ? "Rate limited"
            : "Blocked"
        guard let countdown = resetCountdown(resetsAt: blocked.resetsAt, now: now) else {
            return label
        }
        return "\(label) · \(countdown)"
    }

    // MARK: - Accounts

    /// The caption after the agent name. Locked ×4 and, since EXP-694, as
    /// short as it can be: the bare `<email>` (no "signed in as" prefix, no
    /// " · <plan>" suffix — the row's context already says both) / the bare
    /// plan (pi, which reports a provider instead of an email) / `signed in` /
    /// `signed out`. A missing report is `unknown` — the device never probed,
    /// which is not "signed out".
    public static func accountCaption(_ account: AgentAccount?) -> String {
        guard let account else { return "unknown" }
        guard account.signedIn == true else { return "signed out" }
        if let email = account.email, !email.isEmpty { return email }
        if let plan = account.plan, !plan.isEmpty { return plan }
        return "signed in"
    }

    /// The whole row: `<agent> · <caption>`.
    public static func accountRow(agent: String, account: AgentAccount?) -> String {
        "\(agent) · \(accountCaption(account))"
    }

    // MARK: - Remote login

    /// The sign-in link an `agent_login` device command published, read off
    /// `device_commands.result`.
    ///
    /// The machine completes that command EARLY — the moment the URL (and,
    /// for codex, the device code) is on its screen — with the JSON body
    /// `{"agent":…,"phase":"url","url":…,"code":…}` (desktop
    /// `coding::agent_login::LoginProgress`). Nil for anything else: a failure
    /// sentence, a result from another command kind, or an older build's plain
    /// text, all of which the caller shows verbatim instead. The URL is
    /// scheme-checked here because the only thing a client does with it is
    /// hand it to the system opener.
    public static func parseAgentLoginResult(_ result: String?) -> (url: String, code: String?)? {
        guard let result, let data = jsonData(result) else { return nil }
        guard let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return nil
        }
        guard let raw = object["url"] as? String,
              let url = URL(string: raw),
              let scheme = url.scheme?.lowercased(),
              scheme == "https" || scheme == "http"
        else { return nil }
        let code = (object["code"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        return (raw, code)
    }

    // MARK: - Session join

    /// The usage bar a coding session shows, or nil when it shows none.
    ///
    /// Renders only for a run that is still going (`running` / `in_review`) on
    /// a machine whose report is FRESH and non-empty for the run's own agent —
    /// a finished run's host limits are nobody's business, and stale numbers
    /// beside a live agent read as current ones.
    ///
    /// The devices-row join mirrors `SessionDevicePresentation` exactly: the
    /// stamped `device_id`, preferring the session owner's own row (two users
    /// may see the same machine id through a shared server row).
    public static func sessionUsage(
        session: CodingSessionEntity,
        devices: [DeviceEntity],
        now: Date = Date()
    ) -> SessionAgentUsage? {
        guard session.status == DomainContract.codingSessionStatusRunning
            || session.status == DomainContract.codingSessionStatusInReview
        else { return nil }
        guard let agent = session.agent, !agent.isEmpty else { return nil }
        guard let row = matchedRow(session: session, devices: devices) else { return nil }
        guard let usage = parseMap(row.agentUsage)?[agent] else { return nil }
        guard isFresh(fetchedAt: usage.fetchedAt, now: now) else { return nil }
        guard let windows = usage.windows, !windows.isEmpty else { return nil }
        return SessionAgentUsage(agent: agent, usage: usage)
    }

    private static func matchedRow(
        session: CodingSessionEntity,
        devices: [DeviceEntity]
    ) -> DeviceEntity? {
        guard let deviceId = session.deviceId, !deviceId.isEmpty else { return nil }
        let byId = devices.filter { $0.deviceId == deviceId }
        return byId.first { $0.userId == session.userId } ?? byId.first
    }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/AgentUsagePresentation.kt
package com.exponential.app.domain

import com.exponential.app.data.api.AgentAccount
import com.exponential.app.data.api.AgentUsage
import com.exponential.app.data.api.AgentUsageWindow
import com.exponential.app.data.db.CodingSessionEntity
import com.exponential.app.data.db.DeviceEntity
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull

/** How loud a usage window reads (EXP-484): ≥75 warns, ≥95 is danger. */
enum class AgentUsageSeverity { Normal, Warning, Danger }

/**
 * One rendered usage window (EXP-688) — everything a card shows, decided once
 * in [AgentUsagePresentation.usageGroups] so the four clients cannot drift.
 * [caption] is empty when the card has nothing to add under its bar.
 */
data class UsageCard(
    val key: String,
    val title: String,
    val percent: Int,
    val severity: AgentUsageSeverity,
    val caption: String,
)

/** A titled run of [UsageCard]s — "Current session", "Weekly limits", "Other". */
data class UsageGroup(
    val key: String,
    val title: String,
    val cards: List<UsageCard>,
)

/**
 * EXP-484: how an agent's auth + usage status presents.
 *
 * Hand-mirrored on web (`lib/agent-usage.ts`), iOS
 * (`AgentUsagePresentation.swift`) and the desktop (`ui/src/usage_bar.rs`)
 * against the SAME fixture and the same test names — the strings and the
 * thresholds are the contract, so change all four or none.
 *
 * Everything here is pure and tolerant: the machine's payload is jsonb the
 * server only clamps structurally, so a malformed window is dropped rather
 * than blanking the row, and an unparseable timestamp reads NOT fresh
 * (fail closed — showing a stale percentage as live is the bad direction).
 */
object AgentUsagePresentation {

    /** Usage older than this simply isn't rendered. */
    const val FRESH_WINDOW_MS = 15 * 60_000L

    private val json = Json { ignoreUnknownKeys = true }

    // ── Parsing ──────────────────────────────────────────────────────────────

    /** One agent's `agent_usage` entry as stored/sent; null on anything unusable. */
    fun parseUsage(raw: String?): AgentUsage? =
        usageFrom(parseElement(raw))

    /** The whole `agent_usage` object → per-agent usage; null on anything unusable. */
    fun parseUsageMap(raw: String?): Map<String, AgentUsage>? {
        val root = parseElement(raw) as? JsonObject ?: return null
        return root.mapNotNull { (agent, value) ->
            usageFrom(value)?.let { agent to it }
        }.toMap()
    }

    /** The whole `agent_accounts` object → per-agent sign-in; null on anything unusable. */
    fun parseAccounts(raw: String?): Map<String, AgentAccount>? {
        val root = parseElement(raw) as? JsonObject ?: return null
        return root.mapNotNull { (agent, value) ->
            runCatching { json.decodeFromJsonElement(AgentAccount.serializer(), value) }
                .getOrNull()
                ?.let { agent to it }
        }.toMap()
    }

    private fun parseElement(raw: String?): JsonElement? =
        raw?.takeIf { it.isNotBlank() }?.let {
            runCatching { json.parseToJsonElement(it) }.getOrNull()
        }

    private fun usageFrom(element: JsonElement?): AgentUsage? {
        val obj = element as? JsonObject ?: return null
        return AgentUsage(
            fetchedAt = obj.string("fetchedAt"),
            stale = obj.bool("stale"),
            windows = (obj["windows"] as? JsonArray).orEmpty().mapNotNull(::windowFrom),
        )
    }

    // A window without a key can't be grouped or titled, so it is dropped
    // instead of failing the whole snapshot; the percent is clamped here so
    // every renderer can trust 0-100.
    private fun windowFrom(element: JsonElement): AgentUsageWindow? {
        val obj = element as? JsonObject ?: return null
        val key = obj.string("key")?.takeIf { it.isNotBlank() } ?: return null
        return AgentUsageWindow(
            key = key,
            label = obj.string("label").orEmpty(),
            percent = obj.number("percent").coerceIn(0.0, 100.0),
            resetsAt = obj.string("resetsAt"),
        )
    }

    private fun JsonObject.primitive(name: String): JsonPrimitive? = this[name] as? JsonPrimitive

    private fun JsonObject.string(name: String): String? =
        primitive(name)?.takeIf { it.isString }?.content

    private fun JsonObject.bool(name: String): Boolean =
        primitive(name)?.let { it.booleanOrNull ?: (it.content == "t") } ?: false

    private fun JsonObject.number(name: String): Double =
        primitive(name)?.let { it.doubleOrNull ?: it.content.toDoubleOrNull() } ?: 0.0

    // ── Grouping + severity ──────────────────────────────────────────────────

    /**
     * Every reported window as the cards a Usage surface renders, grouped the
     * way the agent's own app groups them (EXP-688): the current session, the
     * weekly limits (all models, then each per-model window in report order),
     * and everything else (credits, codex's month) under "Other". Empty groups
     * are omitted and the order is fixed.
     *
     * There is no picking any more: the pinned-window concept is gone, so this
     * is the ONLY selection rule and all four clients render the same list.
     */
    fun usageGroups(usage: AgentUsage, nowMs: Long): List<UsageGroup> {
        val session = mutableListOf<UsageCard>()
        val weekly = mutableListOf<UsageCard>()
        val models = mutableListOf<UsageCard>()
        val other = mutableListOf<UsageCard>()
        usage.windows.forEach { window ->
            val card = usageCard(window, nowMs)
            when {
                window.key == WINDOW_SESSION -> session += card
                window.key == WINDOW_WEEKLY -> weekly += card
                window.key.startsWith(MODEL_WINDOW_PREFIX) -> models += card
                else -> other += card
            }
        }
        return buildList {
            if (session.isNotEmpty()) add(UsageGroup(GROUP_SESSION, "Current session", session))
            val weeklyCards = weekly + models
            if (weeklyCards.isNotEmpty()) add(UsageGroup(GROUP_WEEKLY, "Weekly limits", weeklyCards))
            if (other.isNotEmpty()) add(UsageGroup(GROUP_OTHER, "Other", other))
        }
    }

    /**
     * One window as a card. The title names the LIMIT, not the wire key —
     * `Current session` / `All models` / `<Label> only` — and anything the
     * collector doesn't group keeps its own label (`Credits`, `Month`).
     *
     * The caption is the countdown while the window resets, and for an idle
     * session window (0% and no reset — what Claude's own app shows before the
     * first message) says so instead of reading like a dead bar.
     */
    private fun usageCard(window: AgentUsageWindow, nowMs: Long): UsageCard {
        val title = when {
            window.key == WINDOW_SESSION -> "Current session"
            window.key == WINDOW_WEEKLY -> "All models"
            window.key.startsWith(MODEL_WINDOW_PREFIX) -> "${window.label} only"
            else -> window.label
        }
        val caption = resetCountdown(window.resetsAt, nowMs)
            ?: if (window.key == WINDOW_SESSION && window.percent.toInt() == 0) {
                "Starts when a message is sent"
            } else {
                ""
            }
        return UsageCard(
            key = window.key,
            title = title,
            percent = window.percent.toInt(),
            severity = severity(window.percent),
            caption = caption,
        )
    }

    /** The current-session window's wire key. */
    private const val WINDOW_SESSION = "session"

    /** The all-models weekly window's wire key. */
    private const val WINDOW_WEEKLY = "weekly"

    /** Per-model weekly windows are `model:<display>` (`model:fable`). */
    private const val MODEL_WINDOW_PREFIX = "model:"

    const val GROUP_SESSION = "session"
    const val GROUP_WEEKLY = "weekly"
    const val GROUP_OTHER = "other"

    fun severity(percent: Double): AgentUsageSeverity = when {
        percent >= 95.0 -> AgentUsageSeverity.Danger
        percent >= 75.0 -> AgentUsageSeverity.Warning
        else -> AgentUsageSeverity.Normal
    }

    // ── Freshness + countdown ────────────────────────────────────────────────

    /**
     * Whether [fetchedAt] is recent enough to render. FAIL-CLOSED: an absent or
     * unparseable stamp is NOT fresh. A stamp in the future (the machine's
     * clock runs ahead) counts as fresh.
     */
    fun isFresh(fetchedAt: String?, nowMs: Long): Boolean {
        val at = fetchedAt?.let(WireTimestamps::parseEpochMs) ?: return false
        return nowMs - at < FRESH_WINDOW_MS
    }

    /**
     * `resets in 2h 10m` / `resets in 3d 14h` / `resets in 45m` / `resets soon`
     * (under a minute, or already past). Null when the window never resets or
     * the stamp is unreadable. A zero smaller unit is dropped, so an exact
     * boundary reads `resets in 2h`, never `resets in 2h 0m`.
     */
    fun resetCountdown(resetsAt: String?, nowMs: Long): String? {
        val at = resetsAt?.let(WireTimestamps::parseEpochMs) ?: return null
        val minutes = (at - nowMs) / 60_000L
        if (minutes < 1L) return "resets soon"
        val days = minutes / (60L * 24L)
        val hours = (minutes / 60L) % 24L
        return when {
            days > 0L -> if (hours > 0L) "resets in ${days}d ${hours}h" else "resets in ${days}d"
            hours > 0L -> {
                val rest = minutes % 60L
                if (rest > 0L) "resets in ${hours}h ${rest}m" else "resets in ${hours}h"
            }
            else -> "resets in ${minutes}m"
        }
    }

    // ── Account captions ─────────────────────────────────────────────────────

    /**
     * What one agent's sign-in reads as: `signed in as <email> · <plan>`,
     * `signed in as <email>`, the bare plan for an account with no email (pi's
     * `anthropic (oauth)`), `signed in`, `signed out`, or `unknown` when the
     * machine reported nothing for the agent.
     */
    fun accountCaption(account: AgentAccount?): String {
        if (account == null) return "unknown"
        if (!account.signedIn) return "signed out"
        val email = account.email?.takeIf { it.isNotBlank() }
        val plan = account.plan?.takeIf { it.isNotBlank() }
        return when {
            email != null && plan != null -> "signed in as $email · $plan"
            email != null -> "signed in as $email"
            plan != null -> plan
            else -> "signed in"
        }
    }

    /** The whole row: `claude · signed in as danny@yourev.at · max`. */
    fun accountRow(agent: String, account: AgentAccount?): String =
        "$agent · ${accountCaption(account)}"

    // ── Session join ─────────────────────────────────────────────────────────

    /**
     * The usage a LIVE session view renders (its Usage sheet since EXP-688):
     * the host machine's numbers for the agent that run launched with.
     *
     * Nothing renders unless every piece lines up — the run is still working
     * (`running` / `in_review`), it recorded its agent, its host row is the one
     * [resolveSessionDevice] would pick, that row reports fresh usage for the
     * agent, and there is at least one window. A missing piece is normal (old
     * rows, a machine that never reported), so this stays silent rather than
     * showing a placeholder.
     */
    fun sessionUsage(
        session: CodingSessionEntity,
        devices: List<DeviceEntity>,
        nowMs: Long,
    ): AgentUsage? {
        if (session.status != DomainContract.codingSessionStatusRunning &&
            session.status != DomainContract.codingSessionStatusInReview
        ) {
            return null
        }
        val agent = session.agent?.takeIf { it.isNotBlank() } ?: return null
        val deviceId = session.deviceId ?: return null
        val matches = devices.filter { it.deviceId == deviceId }
        val row = matches.firstOrNull { it.userId == session.userId } ?: matches.firstOrNull() ?: return null
        val usage = parseUsageMap(row.agentUsage)?.get(agent) ?: return null
        if (!isFresh(usage.fetchedAt, nowMs)) return null
        return usage.takeIf { it.windows.isNotEmpty() }
    }
}

/**
 * EXP-484: what a completed `agent_login` device command published — the
 * machine's own sign-in URL, plus the device code codex prints (claude has
 * none). The command result is the JSON
 * `{"agent":"codex","phase":"url","url":"https://…","code":"ABCD-EFGHI"}`,
 * capped at 2000 chars server-side.
 */
data class AgentLoginResult(val url: String, val code: String?)

/**
 * Parse a `done` [AgentLoginResult] out of a device command's result string.
 * Null for anything that isn't a login publication — a worktree command's
 * plain-text summary, a truncated payload, a JSON object without a URL — so
 * the caller falls through to the ordinary command caption instead of
 * rendering an empty link.
 */
fun parseAgentLoginResult(result: String?): AgentLoginResult? {
    val trimmed = result?.trim()?.takeIf { it.startsWith("{") } ?: return null
    val obj = runCatching { Json.parseToJsonElement(trimmed) }.getOrNull() as? JsonObject
        ?: return null
    fun text(name: String): String? =
        (obj[name] as? JsonPrimitive)?.takeIf { it.isString }?.content?.takeIf { it.isNotBlank() }
    val url = text("url") ?: return null
    return AgentLoginResult(url = url, code = text("code"))
}
